src/algorithms/time_evolution/simpleupdate3site.jl

```diff
@@ -0,0 +1,523 @@
+    normalize && (l /= norm(l, Inf))
+    return l, q
+end
+function lq_through(
+    M::AbstractTensorMap{T,S,1,4}, ::Nothing; normalize::Bool=true
+) where {T<:Number,S<:ElementarySpace}
+    l, q = rightorth(M; alg=LQpos())
+    @assert isdual(domain(l, 1)) == isdual(codomain(l, 1))
+    normalize && (l /= norm(l, Inf))
+    return l, q
+end
+
+"""
+Given a cluster `Ms`, find all `R`, `L` matrices on each internal bond
+"""
+function _get_allRLs(Ms::Vector{T}) where {T<:AbstractTensorMap}
+    # M1 -- (R1,L1) -- M2 -- (R2,L2) -- M3
+    N = length(Ms)
+    Rs = Vector{AbstractTensorMap}(undef, N - 1)
+    Ls = Vector{AbstractTensorMap}(undef, N - 1)
+    for n in 1:(N - 1)
+        m = N - n + 1
+        _, Rs[n] = qr_through((n == 1) ? nothing : Rs[n - 1], Ms[n]; normalize=true)
+        Ls[m - 1], _ = lq_through(Ms[m], (m == N) ? nothing : Ls[m]; normalize=true)
+    end
+    return Rs, Ls
+end
+
+"""
+Given the tensors `R`, `L` on a bond, construct 
+the projectors `Pa`, `Pb` and the new bond weight `s`
+such that the contraction of `Pa`, `s`, `Pb` is identity when `trunc = notrunc`,
+
+The arrows between `Pa`, `s`, `Pb` are
+```
+    rev = false: - Pa ← s ← Pb -
+                    1 ← s ← 2
+
+    rev = true:  - Pa → s → Pb - 
+                    2 → s → 1
+```
+"""
+function _proj_from_RL(
+    r::AbstractTensorMap{T,S,1,1},
+    l::AbstractTensorMap{T,S,1,1};
+    trunc::TensorKit.TruncationScheme=notrunc(),
+    rev::Bool=false,
+) where {T<:Number,S<:ElementarySpace}
+    rl = r * l
+    @assert isdual(domain(rl, 1)) == isdual(codomain(rl, 1))
+    u, s, vh, ϵ = tsvd!(rl; trunc)
+    sinv = PEPSKit.sdiag_pow(s, -1)
+    Pa, Pb = l * vh' * sinv, sinv * u' * r
+    if rev
+        V = space(Pb, 1)
+        f = isomorphism(flip(V), V)
+        f2 = twist(f, 1)
+        # Pa ← f†) → (f ← s ← f†) → (f ← Pb
+        @tensor begin
+            Pa[-1; -2] := Pa[-1; 1] * f'[1; -2]
+            Pb[-1; -2] := f2[-1; 1] * Pb[1; -2]
+            s[-1; -2] := f2[-1; 1] * s[1; 2] * f'[2; -2]
+        end
+        s = DiagonalTensorMap(permute(s, ((2,), (1,))))
+        @assert all(s.data .>= 0.0)
+    end
+    return Pa, s, Pb, ϵ
+end
+
+"""
+Given a cluster `Ms` and the pre-calculated `R`, `L` bond matrices,
+find all projectors `Pa`, `Pb` and Schmidt weights `wts` on internal bonds.
+"""
+function _get_allprojs(Ms, Rs, Ls, trunc::TensorKit.TruncationScheme, revs::Vector{Bool})
+    N = length(Ms)
+    Pas = Vector{AbstractTensorMap}(undef, N - 1)
+    wts = Vector{DiagonalTensorMap}(undef, N - 1)
+    Pbs = Vector{AbstractTensorMap}(undef, N - 1)
+    # local truncation error on each bond
+    ϵs = zeros(N - 1)
+    for (i, (R, L, rev)) in enumerate(zip(Rs, Ls, revs))
+        trunc2 = if isa(trunc, FixedSpaceTruncation)
+            truncspace(space(Ms[i + 1], 1))
+        else
+            trunc
+        end
+        Pas[i], wts[i], Pbs[i], ϵs[i] = _proj_from_RL(R, L; trunc=trunc2, rev)
+    end
```

Comment:
This kind of construction is actually type-unstable: you instantiate the vectors with abstract types so the compiler no longer has the information about the concrete types of the elements.
Something worth trying is:

```suggestion
    Pas = Vector{AbstractTensorMap}(undef, N - 1)
    wts = Vector{DiagonalTensorMap}(undef, N - 1)
    Pbs = Vector{AbstractTensorMap}(undef, N - 1)
    # local truncation error on each bond
    ϵs = zeros(N - 1)
    Pas_wts_Pbs = map(1:(N-1)) do i
        trunc2 = if isa(trunc, FixedSpaceTruncation)
            truncspace(space(Ms[i + 1], 1))
        else
            trunc
        end
        Pa, wt, Pb, ϵs[i] = _proj_from_RL(Rs[i], Ls[i]; trunc=trunc2, rev)
        return Pa, wt, Pb
    end
    Pas = map(Base.Fix2(getindex, 1), Pas_wts_Pbs)
    wts = map(Base.Fix2(getindex, 2), Pas_wts_Pbs)
    Pbs = map(Base.Fix2(getindex, 3), Pas_wts_Pbs)
```
The alternative would be to try and compute the concrete types of the results